### ISSUE TITLE
Better error message when redefining functions used in desugaring

### ISSUE THREAD

**Issue by chrismshelton** [purescript/purescript#3483]
I'm on the fence as to whether this is actually a problem or not, but it was a frustrating enough experience that I thought it might be worth creating an issue.

Here's some test code that illustrates my issue:
```purescript
module Test where
import Prelude
import Effect (Effect)
import Effect.Console (log)

newtype Buffer = Buffer Unit
newtype UnrelatedThing = UnrelatedThing Unit

bind :: Buffer -> Effect Unit
bind buf = pure unit

newThing :: Effect UnrelatedThing
newThing = do
  u <- log "This is just here so I have a 'bind' somewhere"
  pure (UnrelatedThing u)
```

The errors I recieved were:
```
Warning found:
in module Test
at <module> line 0, column 0 - line 0, column 0

  Shadowed definitions are in scope for value bind from the following open imports:

    import Prelude

  These will be ignored and the local declaration will be used.


See https://github.com/purescript/documentation/blob/master/errors/ScopeShadowing.md for more information,
or to contribute content related to this warning.


Error found:
in module Test
at src/Test.purs line 14, column 3 - line 15, column 3

  Could not match type
               
    Effect Unit
               
  with type
          
    Buffer
          

while checking that type Effect Unit
  is at least as general as type Buffer
while checking that expression log "This is just here so I have a 'bind' somewhere"
  has type Buffer
in value declaration newThing

See https://github.com/purescript/documentation/blob/master/errors/TypesDoNotUnify.md for more information,
or to contribute content related to this error.
```

The warning & error _did_ tell me what was wrong, sort of; but I spent about 2 hours before I put the two things together. Honestly, I am in the habit of ignoring the "shadowed name" warnings while writing new code, and fixing them up at the end - so that's my bad - but still, it was a pretty confusing experience, as I couldn't for the life of me figure out how `newThing` had anything to do in any way with a `Buffer`. Even after reading the warning carefully, I didn't immediately make the connection that the `bind` I was shadowing was the `>>=` bind; I also didn't realize you could redefine bind (which I saw in #3375) - which is pretty cool, but very surprising when you're unaware of it.

To summarize, when you're not aware that you can redefine functions used in desugaring (like `bind`), the resulting errors are extremely confusing. I feel like maybe these should produce different warnings than shadowing other functions (like `reverse` or something); maybe just pointing out that "you're shadowing something the compiler is going to insert all over the place and that might not be what you want"?

**Comment by natefaubion** [comment]
Being able to rebind desugared names is considered a feature, so I'm not sure it would be appropriate to warn in the general case. But maybe it would be possible to add some context to a type error resulting from rebound desugared syntax.
